src/parser/core/modules/__tests__/PrecastStatus.ts

```diff
@@ -0,0 +1,477 @@
+// region CodeUnderTest
+import {PrecastStatus} from '../PrecastStatus'
+// endregion
+// tslint:disable:no-magic-numbers
+// region Mocks
+import {MockedParser} from '__mocks__/Parser'
+import Parser from 'parser/core/Parser'
+import {MockedData} from '__mocks__/Data'
+
+let precastStatus: PrecastStatus
+let parser: Parser
+
+// region MockedData
+import {Ability, AbilityType} from 'fflogs'
+// region MockedEvents
+import {mockApplyBuffEvent, mockApplyBuffStackEvent, mockCastEvent, mockRemoveBuffEvent} from '__mocks__/Events'
+import {Event} from 'events'
+
+let events: Event[] = []
+
+const data = new MockedData()
+
+const mockActionEvent: Ability = {
+	guid: 1,
+	name: 'Ability 1',
+	abilityIcon: '',
+	type: AbilityType.SPECIAL,
+}
+const mockActionData = {
+	id: 1,
+	name: 'Ability 1',
+	icon: '',
+	statusesApplied: ['Status 1'],
+}
+data.mockAction(mockActionData)
+const mockActionStackEvent: Ability = {
+	guid: 2,
+	name: 'Ability 2',
+	abilityIcon: '',
+	type: AbilityType.SPECIAL,
+}
+const mockActionStackData = {
+	id: 2,
+	name: 'Ability 2',
+	icon: '',
+	statusesApplied: ['Status 2'],
+}
+data.mockAction(mockActionStackData)
+
+const mockStatusEvent: Ability = {
+	guid: 1000001,
+	name: 'Status 1',
+	abilityIcon: '',
+	type: AbilityType.SPECIAL,
+}
+const mockStatusData = {
+	id: 1,
+	name: 'Status 1',
+	icon: '',
+}
+data.mockStatus(mockStatusData)
+
+const mockStatusStackEvent: Ability = {
+	guid: 1000002,
+	name: 'Status 2',
+	abilityIcon: '',
+	type: AbilityType.SPECIAL,
+}
+const mockStatusStackData = {
+	id: 2,
+	name: 'Status 2',
+	icon: '',
+	stacksApplied: 3,
+}
+data.mockStatus(mockStatusStackData)
+// endregion
+
+// endregion
+
+// endregion
+
+
+describe('The PrecastStatus module', () => {
+	beforeEach(() => {
+		parser = new MockedParser()
+		Object.defineProperty(parser, 'fight', {value: {start_time: 0}})
+		Object.defineProperty(parser, 'modules', {value: {data}})
+
+		precastStatus = new PrecastStatus(parser)
+		jest.spyOn(data, 'getAction')
+		jest.spyOn(data, 'getActionAppliedByStatus')
+		jest.spyOn(data, 'getStatus')
+	})
+
+	afterEach(() => {
+		jest.clearAllMocks()
+	})
+
+	it('has a normalise method', () => {
```

Comment:
a lot of this seems to be testing the minutae of the implementation details of _how_ PCS works, not _what_ it does.

Assuming an identical end-result, a total rewrite of implementation details should require minimal to no changes to the tests.

Reply:
Sorry, TDD habit from my work code base, we don't usually clean up those sort of things along the development way.  Will feex.